packages/core/src/render3/jit/module.ts

```diff
@@ -19,14 +20,63 @@ import {reflectDependencies} from './util';
 
 const EMPTY_ARRAY: Type<any>[] = [];
 
+interface ModuleQueueItem {
+  moduleType: Type<any>;
+  ngModule: NgModule;
+}
+
+const moduleQueue: ModuleQueueItem[] = [];
+
+/**
+ * Enqueues moduleDef to be checked later to see if scope can be set on its
+ * component declarations.
+ */
+function enqueueModuleForDelayedScoping(moduleType: Type<any>, ngModule: NgModule) {
+  moduleQueue.push({moduleType, ngModule});
+}
+
+/**
+ * Loops over queued module definitions, if a given module definition has all of its
+ * declarations resolved, it dequeues that module definition and sets the scope on
+ * its declarations.
+ */
+export function flushModuleScopingQueueAsMuchAsPossible() {
+  for (let i = moduleQueue.length - 1; i--;) {
+    const {moduleType, ngModule} = moduleQueue[i];
+
+    if (ngModule.declarations && ngModule.declarations.every(isResolvedDeclaration)) {
+      // dequeue
+      moduleQueue.slice(i, 1);
+      setScopeOnDeclaredComponents(moduleType, ngModule);
+    }
+  }
+}
+
+/**
+ * Returns truthy if a declaration has resolved. If the declaration happens to be
+ * an array of declarations, it will recurse to check each declaration in that array
+ * (which may also be arrays).
+ */
+function isResolvedDeclaration(declaration: any[] | Type<any>): boolean {
+  if (Array.isArray(declaration)) {
+    return declaration.every(isResolvedDeclaration);
+  }
+  return typeof declaration !== 'function' || !!resolveForwardRef(declaration);
```

Comment:
```suggestion
  return !!resolveForwardRef(declaration);
```

Declarations are functions themselves (constructor functions). `resolveForwardRef` knows how to deal with itself.